code/ai/aiturret.cpp

```diff
@@ -2380,105 +2361,119 @@ void ai_fire_from_turret(ship *shipp, ship_subsys *ss, int parent_objnum)
 			WeaponMinRange = wip->WeaponMinRange;
 		}
 
-
+		// if beam weapon in nebula and target not tagged, decrease firing range
 		if (wip->wi_flags[Weapon::Info_Flags::Beam]) {
 			if ( !((shipp->tag_left > 0) || (shipp->level2_tag_left > 0)) ) {
-				if (Nebula_sec_range)
-				{
+				if (Nebula_sec_range) {
 					weapon_firing_range *= BEAM_NEBULA_RANGE_REDUCE_FACTOR;
-
-					// *Scale minimum weapon range in nebula    -Et1
-					//Why? - WMC
-					//Commented this out, because it doesn't make sense - WMC
-					//WeaponMinRange *= BEAM_NEBULA_RANGE_REDUCE_FACTOR;
 				}
 			}
 		}
 
-		if(lep != NULL)
-		{
-			if( (dist_to_enemy <= weapon_firing_range) && (dist_to_enemy >= WeaponMinRange) )
-			{
-				if ( wip->wi_flags[Weapon::Info_Flags::Huge] ) {
-					if ( lep->type != OBJ_SHIP ) {
-						continue;
-					}
-					if ( !(Ship_info[Ships[lep->instance].ship_info_index].is_big_or_huge()) ) {
-						continue;
-					}
-				}
+		// Don't try to fire beyond weapon_limit_range (or within min range)
+		if (dist_to_enemy < WeaponMinRange || dist_to_enemy > weapon_firing_range) {
+			// it's possible another weapon is in range, but if not,
+			// we will end up selecting a new target
+			continue;
+		}
 
-				if ( wip->wi_flags[Weapon::Info_Flags::Small_only] ) {
-					if ( (lep->type == OBJ_SHIP) && (Ship_info[Ships[lep->instance].ship_info_index].is_big_or_huge()) ) {
-						continue;
-					}
-				}
+		//	If targeted a small ship and have a huge weapon, don't fire.  But this shouldn't happen, as a small ship should not get selected.
+		if ( wip->wi_flags[Weapon::Info_Flags::Huge] ) {
+			if ( lep->type != OBJ_SHIP ) {
+				tentative_return = true;
+				continue;
+			}
+			if ( !(Ship_info[Ships[lep->instance].ship_info_index].is_big_or_huge()) ) {
+				tentative_return = true;
+				continue;
+			}
+		}
 
-				bool tagged_only = ((wip->wi_flags[Weapon::Info_Flags::Tagged_only]) || (ss->weapons.flags[Ship::Weapon_Flags::Tagged_Only]));
+		// Similar check for small weapons
+		if ( wip->wi_flags[Weapon::Info_Flags::Small_only] ) {
+			if ( lep->type != OBJ_SHIP) {
+				tentative_return = true;
+				continue;
+			}
+			if ( Ship_info[Ships[lep->instance].ship_info_index].is_big_or_huge() ) {
+				tentative_return = true;
+				continue;
+			}
```

Comment:
This changes the logic of `small_only` targeting. In the original, `small_only` weapons _can_ be fired at non-ships, basically the inverse of huge (must be a ship AND big|huge).

Reply:
Ah, good catch.  New commit pushed.

Reply:
And no worries about missing the `fire_down_normals` check.  We all miss something occasionally!